src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/UserAccessScopeMapper.kt

```diff
@@ -0,0 +1,51 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.authorization
+
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.AuthUser
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.DynamicFrameworkContract
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ServiceProvider
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.DynamicFrameworkContractRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ServiceProviderRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.HMPPSAuthService
+import java.lang.RuntimeException
+
+data class UserAccessScope(
+  val serviceProvider: ServiceProvider,
+  val contracts: List<DynamicFrameworkContract>,
+)
+
+@Component
+class UserAccessScopeMapper(
+  private val hmppsAuthService: HMPPSAuthService,
+  private val serviceProviderRepository: ServiceProviderRepository,
+  private val dynamicFrameworkContractRepository: DynamicFrameworkContractRepository,
+) {
+  private val serviceProviderGroupPrefix = "INT_SP_"
+  private val contractGroupPrefix = "INT_CG_"
+
+  // get the user access scope for a given service provider user
+  fun forServiceProviderUser(user: AuthUser): UserAccessScope {
+    val groups = hmppsAuthService.getUserGroups(user)
+      ?: throw RuntimeException("pp or non existent user")
+
+    val serviceProviderGroups = groups
+      .filter { it.startsWith(serviceProviderGroupPrefix) }
+      .map { it.removePrefix(serviceProviderGroupPrefix) }
+
+    if (serviceProviderGroups.size != 1) {
+      throw RuntimeException("there must only be one service provider group per user")
+    }
+
+    val contractGroups = groups
+      .filter { it.startsWith(contractGroupPrefix) }
+      .map { it.removePrefix(contractGroupPrefix) }
+
+    val serviceProvider = serviceProviderRepository.findByIdOrNull(serviceProviderGroups[0])
+      ?: throw RuntimeException("user does not exist in our database")
```

Comment:
this raised a few questions in my head about under what circumstances users are persisted. 

currently when:

 - an appointment is created by an SP user
 - an action plan is created or submitted by an SP user
 - an EOSR is created or submitted by an SP user
 - a referral is assigned to a caseworker by an SP user
 - a referral is assigned to an SP user
 - a referral end is requested by a PP user
 - a referral is created or sent by a PP user
 
 are there circumstances where a user may be trying to access a referral but has never been saved in the db by one of the actions listed above? should we persist all users when they log in to the service too?